src/shutdown.rs

```diff
@@ -0,0 +1,253 @@
+    /// or not the source shutdown gracefully.
+    // TODO: The timing and reporting logic is very similar to the logic in
+    // `RunningTopology::stop()`. Once `RunningTopology::stop()` has been updated to utilize the
+    // ShutdownCoordinator, see if some of this logic can be de-duped.
+    pub fn shutdown_source_end<'a>(
+        &mut self,
+        rt: &mut runtime::Runtime,
+        name: String,
+        deadline: Instant,
+    ) -> bool {
+        let name2 = name.clone();
+        let name3 = name.clone();
+        let shutdown_complete_tripwire = self.shutdown_complete_tripwires.remove(&name).unwrap();
+        let shutdown_force_trigger = self.shutdown_force_triggers.remove(&name).unwrap();
+
+        let success = shutdown_complete_tripwire.map(move |_| {
+            info!("Source \"{}\" shut down successfully", name);
+        });
+        let timeout = timer::Delay::new(deadline)
+            .map(move |_| {
+                error!(
+                    "Source '{}' failed to shutdown before deadline. Forcing shutdown.",
+                    name2,
+                );
+            })
+            .map_err(|err| panic!("Timer error: {:?}", err));
+        let reporter = timer::Interval::new_interval(Duration::from_secs(5))
+            .inspect(move |_| {
+                let time_remaining = if deadline > Instant::now() {
+                    format!(
+                        "{} seconds remaining",
+                        (deadline - Instant::now()).as_secs()
+                    )
+                } else {
+                    "overdue".to_string()
+                };
+
+                info!(
+                    "Still waiting on source \"{}\" to shut down. {}",
+                    name3, time_remaining,
+                );
+            })
+            .filter(|_| false) // Run indefinitely without emitting items
+            .into_future()
+            .map(|_| ())
+            .map_err(|(err, _)| panic!("Timer error: {:?}", err));
+
+        let union = future::select_all::<Vec<Box<dyn Future<Item = (), Error = ()> + Send>>>(vec![
+            Box::new(success),
+            Box::new(timeout),
+            Box::new(reporter),
+        ]);
+
+        let (_, index, _) = rt.block_on(union).ok().unwrap();
+        let success = if index == 0 {
+            true
+        } else if index == 1 {
+            false
+        } else {
+            panic!(
+                "Neither success nor timeout future finished.  Index finished: {}",
+                index
+            );
+        };
+        if success {
+            // TODO Shouldn't need to cancel here, we should be able to call
+            // `shutdown_force_trigger.disable()` instead.  We can't yet though because
+            // some topology tests rely on notifying this trigger to mask failures of the
+            // underlying 'source_task'.
+            shutdown_force_trigger.cancel();
```

Comment:
This is interesting and was the cause of some of the failing topology integration tests that my previous version was encountering.  With this line changed to call `.ignore()` instead of `.cancel()` we get a failure in two topology tests: `topology_swap_source` and `topology_remove_one_source`, which notably are the only tests that ever result in `RunningTopology::shutdown_source_{begin,end}` being called.  They fail in `shutdown_source_end` at the line `self.source_tasks.remove(name).wait().unwrap();`, because the `wait()` call is returning an `Err` not an `Ok`.  Given that these tests are hitting this when removing a Source out from under a running topology, I suspect that the `source_task` for the MockSource used by the topology tests has *always* been returning an error, but `select`-ing on the `shutdown_tripwire` has been masking that failure.

Reply:
Whoa! Interesting.

Reply:
I wonder if this has to do with the fact that the mock source ignores the shutdown signal and just drops it before the future for that source is even polled?

Reply:
I decided to start my day today by diving into this and I figured out the issue.  The problem was with `stream_cancel::{Trigger, Tripwire}` (which are wrappers around `oneshot::{Sender, Receiver}`).  Turns out if the `Trigger` is dropped it causes the `Tripwire` future to resolve with an error, \*even if\* `disable()` was already called on the `Trigger`.  I fixed this by setting using `or_else` with `futures::empty` so that the `Tripwire` Future will never resolve unless the corresponding `Trigger` is notified via `cancel()`, even if that `Trigger` gets dropped.